docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -201,44 +201,117 @@ The ShootState synchronization controller will become part of the gardenlet. It
+* `gardenlet` cannot connect to the seed's `kube-apiserver`. In this case `gardenlet` knows if migration has been triggered, but it will not start shoot migration or reconciliation as it will first check the seed conditions and try to update the `Cluster` resource, both of which will fail. Extension controllers could still be able to connect to the seed's `kube-apiserver` (if they are not running where `gardenlet` is running), and similarly to the previous case, they could still attempt to reconcile their resources.
+* The seed components (`etcd-druid`, extension controllers, etc) cannot connect to the seed's `kube-apiserver`. In this case extension controllers would not be able to reconcile their resources as they cannot fetch them from the seed's `kube-apiserver`. When the connection to the `kube-apiserver` comes back, the controllers might be stuck in a retry loop from a previous reconciliation, or the resources could still be annotated with `gardener.cloud/operation=reconcile`. This could lead to a race condition depending on who manages to `update` or `get` the resources first. If `gardenlet` manages to update the resources before they are read by the extension controllers, they would be properly updated with `gardener.cloud/operation=migrate`. Otherwise, they would be reconciled as usual.
+
+The safety mechanism is based on "cluster leases". Essentially, extension controllers must check if the seed has a valid "lease" for the shoot before attempting a reconciliation. The lease is just an additional timestamp field in the `Cluster` resource:
+
+```go
+// ClusterSpec is the spec for a Cluster resource.
+type ClusterSpec struct {
+	...
+	// LeaseExpiration indicates when the lease of this cluster will expire
+	LeaseExpiration metav1.MicroTime `json:"leaseExpiration"`
+}
+```
 
-When case 3 happens it means that the shoot has been migrated and the backuprestore container in the  __Destination Seed__ has started using the Shoot's backup entry. The backuprestore on the __Source Seed__ should be configured to shut itself and the etcd container down once it sees the newer timestamp. Shutting down the etcd will cause the kubernetes control plane components to go into crashloop backoff and the MCM will not be able to do anything as it will not be able to list nodes (this has to be verified with MCM).
+On a regular basis, e.g. every 2 minutes, and at each shoot reconciliation, `gardenlet` fetches the shoot from the Garden `kube-apiserver`, and if successful and the control plane migration hasn't been triggered it updates the `Cluster` resource with a renewed timestamp, e.g. 2 minutes from now. The extension controllers fetch this resource before each reconciliation and check if the lease has expired. If it hasn't, they reconcile as usual. If the lease has expired, they skip the reconciliation, since either the control plane migration has been triggered or `gardenlet` cannot reliably determine if the seed is still the owner.
```

Comment:
Apart from your discussion, that's not bullet-proof either, but what is? If the extension controller does its job, it may have checked before running the reconcile, but when the cluster gets migrated in the middle/the lease expires, it won't recheck and hence can update the state when it has long lost the mandate to do so.

We can only try to reduce the likelihood/impact, e.g. by something alike 2PC where the extensions check-and-set that they are working now. While any extension is "busy" with a shoot, a migration may be requested/lease may be scheduled to expire, but this should only block any other/new reconciliation from kicking-off, but the process of migration/replicating the final state back to the garden cluster/running the final backup, only happens after said reconciliation completes (or another timeout lapses - that's not the 2m timeout for the lease as such). Maybe that's too complicated... but would that be an airplane, space rocket, or life-support system, we'd probably do that or something similar to ensure we lose no state?

Reply:
> Apart from your discussion, that's not bullet-proof either, but what is? If the extension controller does its job, it may have checked before running the reconcile, but when the cluster gets migrated in the middle/the lease expires, it won't recheck and hence can update the state when it has long lost the mandate to do so.

This is why we wait in step 11. The restore will not continue unless it's 100% sure the lease had expired. Running reconciliations won't be prevented, but they would have to be really long to cause trouble. Also in this case, a simpler way to deal with it (in the vast majority of cases) would be a longer wait in step 11.

> by something alike 2PC ...

I would first try hard to avoid something as sophisticated, if at all possible. We should think about it, discuss it in a meeting, and extend the document accordingly.

Reply:
In the latest proposal there are no cluster leases anymore, so this discussion is no longer relevant.